yarn-project/acir-simulator/src/public/executor.ts

```diff
@@ -0,0 +1,134 @@
+import { AztecAddress, CallContext, EthAddress, Fr, FunctionData, TxRequest } from '@aztec/circuits.js';
+import { padArrayEnd } from '@aztec/foundation/collection';
+import { createDebugLogger } from '@aztec/foundation/log';
+import { select_return_flattened as selectPublicWitnessFlattened } from '@noir-lang/noir_util_wasm';
+import { acvm, frToAztecAddress, frToSelector, fromACVMField, toACVMField, toACVMWitness } from '../acvm/index.js';
+import { PublicContractsDB, PublicStateDB } from './db.js';
+import { PublicExecution, PublicExecutionResult, isPublicExecution } from './execution.js';
+import { StateActionsCollector } from './state_actions.js';
+
+// Copied from crate::abi at noir-contracts/src/contracts/noir-aztec3/src/abi.nr
+const NOIR_MAX_RETURN_VALUES = 4;
+
+export class PublicExecutor {
+  constructor(
+    public readonly stateDb: PublicStateDB,
+    public readonly contractsDb: PublicContractsDB,
+
+    private log = createDebugLogger('aztec:simulator:public-executor'),
+  ) {}
+
+  public async execute(request: TxRequest): Promise<PublicExecutionResult>;
```

Comment:
Wouldn't it be clearer if the Executor just handles TxRequests and creates Executions with them? and then executions can create nested executions (similar to the simulator=>execution relationship of the private calls). I'm not sure though :D

Reply:
I wanted the execution object to be as dumb as possible, so it just carries the information of the execution itself, while keeping the logic in the executor. Maybe the catch is keeping this `execute` method simpler, to just handle `executions`, and have a separate `createExecutionFromTxRequest` as you suggest? I'll think about it!

Reply:
The `createExecutionFromTxRequest` approach sounds good to me! however I was just being picky about readability, whatever you think is best :D